src/compiler/declarationEmitter.ts

```diff
@@ -1594,7 +1594,12 @@ namespace ts {
                 emitTypeOfVariableDeclarationFromTypeLiteral(node);
             }
             else if (!hasModifier(node.parent, ModifierFlags.Private)) {
-                writeTypeOfDeclaration(node, node.type, getParameterDeclarationTypeVisibilityError);
+                // use the checker's type, not the declared type,
+                // for optional parameters and initialized ones that aren't a parameter property
+                const typeShouldAddUndefined = resolver.isOptionalParameter(node) ||

```

Comment:
Could this be moved into `writeTypeOfDeclaration`. Seems like you're having to do this twice.